Lib/ntpath.py

```diff
@@ -641,22 +639,73 @@ def commonpath(paths):
         raise
 
 
-# determine if two files are in fact the same file
 try:
-    # GetFinalPathNameByHandle is available starting with Windows 6.0.
-    # Windows XP and non-Windows OS'es will mock _getfinalpathname.
-    if sys.getwindowsversion()[:2] >= (6, 0):
-        from nt import _getfinalpathname
-    else:
-        raise ImportError
-except (AttributeError, ImportError):
-    # On Windows XP and earlier, two files are the same if their absolute
-    # pathnames are the same.
-    # Non-Windows operating systems fake this method with an XP
-    # approximation.
-    def _getfinalpathname(f):
-        return normcase(abspath(f))
+    from nt import _getfinalpathname
+
+    # determine if two files are in fact the same file
+    def realpath(filename):
+        filename = os.fspath(filename)
+        extended_path_prefix = '\\\\?\\' if isinstance(filename, str) else b'\\\\?\\'
+        is_extended_path = filename.startswith(extended_path_prefix)
+        unresolved = filename if is_extended_path else abspath(filename)
+        resolved_parts = []
+        while True:
+            try:
+                resolved_parts.append(_getfinalpathname(unresolved))
+                break
+            except OSError:
+                unresolved, tail = split(unresolved)
+                if not tail:
+                    resolved_parts.append(unresolved)
+                    break
+                resolved_parts.append(tail)
+        resolved = join(*reversed(resolved_parts))
+        # try to convert extended path to normal if
+        # initial path did not use \\?\ prefix and result uses it
+        if not is_extended_path and resolved.startswith(extended_path_prefix):
+            resolved = _extended_to_normal(resolved)
+        return resolved
+
+    def _extended_to_normal(path):
+        drive, rest = splitdrive(path)
+        if not rest:
+            return path
+        if isinstance(path, str):
+            sep = '\\'
+            colon = ':'
+            letters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
+            unc_dev = '\\\\?\\UNC'
+        else:
+            sep = b'\\'
+            colon = b':'
+            letters = b'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
+            unc_dev = b'\\\\?\\UNC'
+        drive = drive.upper()
+        if drive == unc_dev:
+            # UNC path with \\?\ prefix - drop prefix
+            # 7 is len('\\?\UNC')
+            normal_path = sep + path[7:]
+        elif len(drive) == 6 and drive[4:5] in letters and drive[5:6] == colon:
+            # extended path with \\?\ prefix
+            # 4 is len('\\?\')
+            normal_path = path[4:]
+        else:
+            # not a UNC or drive-letter path
+            # return path as-is
+            return path
 
+        if 0 < len(normal_path) < 260 and normal_path == abspath(normal_path):
```

Comment:
We don't need to check for a zero-length path in this version of the function. That was a kludge in the rough draft. Do you think it would be worthwhile to define  `MAX_PATH` in this module?

Reply:
dropped '0 <' check in the latest rev. Do you mean to expose `MAX_PATH` as a documented module property? To me it seems reasonable if it meets the bar.